// vtex/lib/modules/support/login.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const chalk_1 = __importDefault(require("chalk"));
const enquirer_1 = __importDefault(require("enquirer"));
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const ramda_1 = require("ramda");
const conf_1 = require("../../conf");
const env = __importStar(require("../../env"));
const logger_1 = __importDefault(require("../../logger"));
const getAvailableRoles = async (region, token, supportedAccount) => {
    const response = await axios_1.default.get(`http://support-authority.vtex.${region}.vtex.io/${conf_1.getAccount()}/${conf_1.getWorkspace()}/${supportedAccount}/roles`, {
        headers: {
            Authorization: token,
            'X-Vtex-Original-Credential': token,
        },
    });
    return response.data;
};
const promptRoles = async (roles) => {
    const cancel = 'Cancel';
    const chosen = ramda_1.prop('role', await enquirer_1.default.prompt({
        name: 'role',
        message: 'Which role do you want to assume?',
        type: 'select',
        choices: [...roles, cancel],
    }));
    if (chosen === cancel) {
        logger_1.default.info('Bye! o/');
        return process.exit();
    }
    return chosen;
};
const loginAsRole = async (region, token, supportedAccount, role) => {
    const response = await axios_1.default.get(`http://support-authority.vtex.${region}.vtex.io/${conf_1.getAccount()}/${conf_1.getWorkspace()}/${supportedAccount}/login/${role}`, {
        headers: {
            Authorization: token,
            'X-Vtex-Original-Credential': token,
        },
    });
    return response.data;
};
const assertToken = (raw) => {
    if (!jsonwebtoken_1.default.decode(raw)) {
        throw Error(`Could not validate new token! token = '${raw}'`);
    }
};
const saveSupportCredentials = (account, token) => {
    conf_1.saveAccount(account);
    conf_1.saveWorkspace('master');
    conf_1.saveToken(token);
};
exports.default = async (account) => {
    if (!account) {
        logger_1.default.error(`Please specify the account that will receive support. type vtex --help for more information.`);
        return;
    }
    const actualToken = conf_1.getToken();
    const region = env.region();
    try {
        const roles = await getAvailableRoles(region, actualToken, account);
        if (roles.length === 0) {
            logger_1.default.error('No support roles available for this account.');
            return;
        }
        const role = await promptRoles(roles);
        const newToken = await loginAsRole(region, actualToken, account, role);
        assertToken(newToken);
        saveSupportCredentials(account, newToken);
        logger_1.default.info(`Logged into ${chalk_1.default.blue(account)} with role ${role}!`);
    }
    catch (err) {
        if (err.message) {
            logger_1.default.error(err.message);
            if (err.response && err.response.status === 404) {
                logger_1.default.info('Make sure vtex.support-authority is installed in your workspace.');
            }
            return;
        }
        logger_1.default.error(err);
    }
};
