// vtex/lib/modules/auth/login.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const chalk_1 = __importDefault(require("chalk"));
const child_process_1 = __importDefault(require("child_process"));
const enquirer_1 = __importDefault(require("enquirer"));
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const opn_1 = __importDefault(require("opn"));
const path_1 = require("path");
const ramda_1 = require("ramda");
const randomstring_1 = __importDefault(require("randomstring"));
const conf = __importStar(require("../../conf"));
const env_1 = require("../../env");
const logger_1 = __importDefault(require("../../logger"));
const sse_1 = require("../../sse");
const prompts_1 = require("../prompts");
const [cachedAccount, cachedLogin, cachedWorkspace] = [conf.getAccount(), conf.getLogin(), conf.getWorkspace()];
const details = cachedAccount && `${chalk_1.default.green(cachedLogin)} @ ${chalk_1.default.green(cachedAccount)} / ${chalk_1.default.green(cachedWorkspace)}`;
const oldLoginUrls = (workspace, state) => {
    const returnUrl = `/_v/private/auth-server/v1/callback?workspace=${workspace}&state=${state}`;
    const url = `/_v/private/auth-server/v1/login/?workspace=${workspace}`;
    return [url, returnUrl];
};
const newLoginUrls = (workspace, state) => {
    const returnUrl = `/_v/private/auth-server/v1/callback?workspace=${workspace}&state=${state}`;
    const url = `/_v/segment/admin-login/v1/login?workspace=${workspace}`;
    return [url, returnUrl];
};
const getLoginUrl = async (account, workspace, state) => {
    const baseUrl = `https://${account}${env_1.clusterIdDomainInfix()}.${env_1.publicEndpoint()}`;
    let [url, returnUrl] = newLoginUrls(workspace, state);
    try {
        const response = await axios_1.default.get(`${baseUrl}${url}`);
        if (!response.data.match(/vtex\.admin-login/)) {
            throw new Error('Unexpected response from admin-login');
        }
    }
    catch (e) {
        const oldUrls = oldLoginUrls(workspace, state);
        url = oldUrls[0];
        returnUrl = oldUrls[1];
    }
    const fullReturnUrl = baseUrl + returnUrl;
    const returnUrlEncoded = encodeURIComponent(returnUrl);
    return [`${baseUrl}${url}&returnUrl=${returnUrlEncoded}`, fullReturnUrl];
};
const startUserAuth = async (account, workspace) => {
    const state = randomstring_1.default.generate();
    const [url, fullReturnUrl] = await getLoginUrl(account, workspace, state);
    opn_1.default(url, { wait: false });
    return sse_1.onAuth(account, workspace, state, fullReturnUrl);
};
const promptUsePrevious = () => prompts_1.promptConfirm(`Do you want to use the previous login details? (${details})`);
const promptAccount = async (promptPreviousAcc) => {
    if (promptPreviousAcc) {
        const confirm = await prompts_1.promptConfirm(`Use previous account? (${chalk_1.default.blue(cachedAccount)})`);
        if (confirm) {
            return cachedAccount;
        }
    }
    const account = ramda_1.prop('account', await enquirer_1.default.prompt({
        type: 'input',
        result: s => s.trim(),
        message: 'Account:',
        name: 'account',
        validate: s => /^\s*[\w-]+\s*$/.test(s) || 'Please enter a valid account.',
    }));
    return account;
};
exports.saveCredentials = (login, account, token, workspace) => {
    conf.saveLogin(login);
    conf.saveAccount(account);
    conf.saveToken(token);
    conf.saveAccountToken(account, token);
    conf.saveWorkspace(workspace);
};
const authAndSave = async (account, workspace, optionWorkspace) => {
    const [token, returnUrl] = await startUserAuth(account, optionWorkspace ? workspace : 'master');
    const decodedToken = jsonwebtoken_1.default.decode(token);
    const login = decodedToken.sub;
    exports.saveCredentials(login, account, token, workspace);
    return { login, token, returnUrl };
};
const closeChromeTabIfMac = (returnUrl) => {
    if (process.platform === 'darwin') {
        const cp = child_process_1.default.spawn('osascript', [path_1.join(__dirname, '../../../scripts/closeChrome.scpt'), returnUrl], {
            stdio: 'ignore',
            detached: true,
        });
        cp.unref();
    }
};
exports.default = async (options) => {
    const defaultArgumentAccount = options && options._ && options._[0];
    const optionAccount = options ? options.a || options.account || defaultArgumentAccount : null;
    const optionWorkspace = options ? options.w || options.workspace : null;
    const usePrevious = !(optionAccount || optionWorkspace) && details && (await promptUsePrevious());
    const account = optionAccount || (usePrevious && cachedAccount) || (await promptAccount(cachedAccount && optionWorkspace));
    const workspace = optionWorkspace || (usePrevious && cachedWorkspace) || 'master';
    try {
        const { login, token, returnUrl } = await authAndSave(account, workspace, optionWorkspace);
        logger_1.default.debug('Login successful', login, account, token, workspace);
        logger_1.default.info(`Logged into ${chalk_1.default.blue(account)} as ${chalk_1.default.green(login)} at workspace ${chalk_1.default.green(workspace)}`);
        closeChromeTabIfMac(returnUrl);
    }
    catch (err) {
        if (err.statusCode === 404) {
            logger_1.default.error('Account/Workspace not found');
        }
        else {
            throw err;
        }
    }
};
