// vtex/lib/clients/Builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const api_1 = require("@vtex/api");
const builderBaseRoute = `/_v/builder/0`;
const routes = {
    tsConfig: `${builderBaseRoute}/tsconfig`,
    typings: `${builderBaseRoute}/typings`,
    availability: (app) => `${routes.builder}/availability/${app}`,
    builder: builderBaseRoute,
    clean: (app) => `${routes.builder}/clean/${app}`,
    link: (app) => `${routes.builder}/link/${app}`,
    pinnedDependencies: () => `${routes.builder}/pinneddeps`,
    publish: (app) => `${routes.builder}/publish/${app}`,
    relink: (app) => `${routes.builder}/relink/${app}`,
    test: (app) => `${routes.builder}/test/${app}`,
};
class Builder extends api_1.AppClient {
    constructor(ioContext, opts) {
        super('vtex.builder-hub@0.x', ioContext, opts);
        this.availability = async (app, hintIndex) => {
            const stickyHint = hintIndex === undefined || hintIndex === null
                ? `request:${this.context.account}:${this.context.workspace}:${app}`
                : `request:${this.context.account}:${this.context.workspace}:${app}:${hintIndex}`;
            const headers = {
                'Content-Type': 'application/json',
                'x-vtex-sticky-host': stickyHint,
            };
            const metric = 'bh-availability';
            const { data: { availability }, headers: { 'x-vtex-sticky-host': host }, } = await this.http.getRaw(routes.availability(app), { headers, metric, cacheable: api_1.CacheType.None });
            const { hostname, score } = availability;
            return { host, hostname, score };
        };
        this.clean = (app) => {
            const headers = {
                'Content-Type': 'application/json',
                ...(this.stickyHost && { 'x-vtex-sticky-host': this.stickyHost }),
            };
            const metric = 'bh-clean';
            return this.http.post(routes.clean(app), { headers, metric });
        };
        this.getPinnedDependencies = () => {
            return this.http.get(routes.pinnedDependencies());
        };
        this.publishApp = (app, zipFile, stickyOptions = { sticky: true }, params = {}) => {
            return this.sendZipFile(routes.publish(app), app, zipFile, stickyOptions, params);
        };
        this.testApp = (app, zipFile, stickyOptions = { sticky: true }, params = {}) => {
            return this.sendZipFile(routes.test(app), app, zipFile, stickyOptions, params);
        };
        this.linkApp = (app, zipFile, stickyOptions = { sticky: true }, params = {}) => {
            return this.sendZipFile(routes.link(app), app, zipFile, stickyOptions, params);
        };
        this.relinkApp = (app, changes, params = {}) => {
            const headers = {
                'Content-Type': 'application/json',
                ...(this.stickyHost && { 'x-vtex-sticky-host': this.stickyHost }),
            };
            const metric = 'bh-relink';
            return this.http.put(routes.relink(app), changes, { headers, metric, params });
        };
        this.builderHubTsConfig = () => {
            return this.http.get(routes.tsConfig);
        };
        this.typingsInfo = async () => {
            const res = await this.http.get(routes.typings);
            return res.typingsInfo;
        };
        this.sendZipFile = async (route, app, zipFile, { tag, sticky, stickyHint } = {}, requestParams = {}) => {
            const hint = stickyHint || `request:${this.context.account}:${this.context.workspace}:${app}`;
            const metric = 'bh-zip-send';
            const params = tag ? { ...requestParams, tag } : requestParams;
            const { data, headers: { 'x-vtex-sticky-host': host }, } = await this.http.postRaw(route, zipFile, {
                headers: {
                    'Content-length': zipFile.byteLength,
                    'Content-Type': 'application/octet-stream',
                    ...(sticky && { 'x-vtex-sticky-host': this.stickyHost || hint }),
                },
                metric,
                params,
            });
            this.stickyHost = host;
            return data;
        };
    }
}
exports.Builder = Builder;
